### PR TITLE
Enhance reliablity on Factory Reset Notification for cloud CPE sync

### PR DESCRIPTION
**Pull request by sadhyama**

### CODE REVIEW

**Comment by schmidtw** [comment]
The changes look good to me.

File: source/app/libpd.c [hunk 1]
```diff
@@ -171,14 +179,41 @@ static void parodus_receive()
...
+
+				if(sourceService != NULL && sourceApplication != NULL && strcmp(sourceService,"parodus")== 0 && strcmp(sourceApplication,"cloud-status")== 0)
+				{
+					WalInfo("cloud-status Retrieve response received from parodus : %s len %lu\n",(char *)wrp_msg->u.crud.payload, strlen(wrp_msg->u.crud.payload) );
+
+					parsePayloadForStatus(wrp_msg->u.crud.payload, &status);
+					if(status !=NULL)
+					{
+						//set this as global conn status. adding lock before update it.
+						pthread_mutex_lock (&cloud_mut);
```
Comment (shilpa24balaji) [inline source/app/libpd.c]: Add lock inside the setter and getter functions for global cloud_status

File: source/app/libpd.c [hunk 2]
```diff
@@ -216,6 +251,111 @@ void parodus_receive_wait()
...
+			}
+
+			if(dest !=NULL)
+			{
+				snprintf(dest, MAX_STR_LENGTH, "mac:%s/parodus/cloud-status", device_mac);
+				req_wrp_msg->u.crud.dest = dest;
+				WalPrint("req_wrp_msg->u.crud.dest is %s\n", req_wrp_msg->u.crud.dest);
+			}
+
+			req_wrp_msg->u.crud.transaction_uuid = strdup("bd4ad2d1-5c9c-486f-8e25-52c242b38");
```
Comment (shilpa24balaji) [inline source/app/libpd.c]: Why is this transaction id hardcoded?

File: source/app/libpd.c (hunk 1 above)
Comment (sadhyama) [reply source/app/libpd.c]: Incorporated.

File: source/app/libpd.c (hunk 2 above)
Comment (sadhyama) [reply source/app/libpd.c]: For CRUD requests, user should pass transaction_id as server is not generating its own and it is returning empty payload response. 

To handle this, added a method to generate random transaction_uuid for every cloud-status retrieve request.

File: source/app/libpd.c [hunk 3]
```diff
@@ -344,6 +501,66 @@ static void get_parodus_url(char **parodus_url, char **client_url)
...
+	}
+}
+
+static void generate_trans_uuid(char **transID)
+{
+	FILE *file = NULL;
+	char buffer [ 64 ] = { 0 };
+	char *command = NULL;
+
+	command = strdup("uuidgen -r");
```
Comment (shilpa24balaji) [inline source/app/libpd.c]: Use C API https://linux.die.net/man/3/uuid_generate

File: source/app/libpd.c [hunk 4]
```diff
@@ -216,6 +251,128 @@ void parodus_receive_wait()
...
+
+				//waiting to get response from parodus. add lock here while reading
+				pthread_mutex_lock (&cloud_mut);
+				WalPrint("mutex lock in consumer thread\n");
+				WalPrint("Before pthread cond wait in consumer thread\n");
+				pthread_cond_wait(&cloud_con, &cloud_mut);
+				pthread_mutex_unlock (&cloud_mut);
+				WalPrint("mutex unlock in consumer thread after cond wait\n");
+
+				if ((get_global_cloud_status() !=NULL) && (strcmp(get_global_cloud_status(), CLOUD_STATUS_ONLINE) == 0))
```
Comment (shilpa24balaji) [inline source/app/libpd.c]: Dont call get_global_cloud_status() multiple times instead use local variable.

File: source/app/libpd.c [hunk 5]
```diff
@@ -171,14 +180,40 @@ static void parodus_receive()
...
+					parsePayloadForStatus(wrp_msg->u.crud.payload, &status);
+					if(status !=NULL)
+					{
+						//set this as global conn status. add lock before update it.
+						set_global_cloud_status(status);
+						WalInfo("set cloud-status value as %s\n", status);
+
+						pthread_mutex_lock (&cloud_mut);
+						WalPrint("mutex lock in producer thread\n");
+						pthread_cond_signal(&cloud_con);
```
Comment (shilpa24balaji) [inline source/app/libpd.c]: I feel it is better to move the pthread_cond_signal and pthread_cond_wait also into the getter and setter functions so it is better abstraction.

File: source/app/libpd.c [hunk 6]
```diff
@@ -216,6 +251,128 @@ void parodus_receive_wait()
...
+					rv = 1;
+					free(get_global_cloud_status());
+					break;
+				}
+				else
+				{
+					WalError("cloud_status is not online, sending RETRIEVE request again and retrying\n" );
+					WalInfo("cloud status retry backoffRetryTime '%d' seconds\n", backoffRetryTime);
+					sleep(backoffRetryTime);
+					c++;
```
Comment (shilpa24balaji) [inline source/app/libpd.c]: c here in backoffRetryTime should be maxed to 9 i.e. ~511s and not keep incrementing

File: source/app/libpd.c [hunk 7]
```diff
@@ -216,6 +251,128 @@ void parodus_receive_wait()
...
+
+			if(dest !=NULL)
+			{
+				snprintf(dest, MAX_STR_LENGTH, "mac:%s/parodus/cloud-status", device_mac);
+				req_wrp_msg->u.crud.dest = dest;
+				WalPrint("req_wrp_msg->u.crud.dest is %s\n", req_wrp_msg->u.crud.dest);
+			}
+
+			generate_trans_uuid(&transaction_uuid);
+			if(transaction_uuid !=NULL)
```
Comment (shilpa24balaji) [inline source/app/libpd.c]: This transaction_uuid should change in every backoff retry attempt

File: source/app/libpd.c [hunk 8]
```diff
@@ -216,6 +246,148 @@ void parodus_receive_wait()
         WalPrint ("End of parodus_upstream\n");
 }
 
+char *get_global_cloud_status()
```
Comment (shilpa24balaji) [inline source/app/libpd.c]: Add comment that combining get with lock and wait so not normal getter func

File: source/app/libpd.c [hunk 9]
```diff
@@ -344,6 +516,51 @@ static void get_parodus_url(char **parodus_url, char **client_url)
...
+		}
+		else
+		{
+			WalError("Failed to get cloudStatus from payload\n");
+		}
+		cJSON_Delete(json);
+	}
+}
+
+static void generate_trans_uuid(char **transID)
```
Comment (shilpa24balaji) [inline source/app/libpd.c]: Instead of double pointer argument, return the new string malloc pointer or NULL

**Pull request merged**